Redox Cross Compilation

jackpot51 opened the pull request:
I will admit - there are things here that I wish I did not have to do. This completes the ability to create a cross compiler from the rust repository for `x86_64-unknown-redox`. I will document this PR with inline comments explaining some things.

[View this gist to see how a cross compiler is built](https://gist.github.com/jackpot51/6680ad973986e84d69c79854249f2b7e)

Prior discussion of a smaller change is here: https://github.com/rust-lang/rust/pull/38366

rust-highfive commented:
r? @sfackler

(rust_highfive has picked a reviewer for you, use r? to override)

jackpot51 commented on mk/cfg/x86_64-unknown-redox.mk:
```diff
@@ -0,0 +1 @@
+# rustbuild-only target
```
The old build system is complicated to use, so there is no support for using it.

jackpot51 commented on src/liballoc_jemalloc/build.rs:
```diff
@@ -36,7 +36,8 @@ fn main() {
     if target.contains("rumprun") || target.contains("bitrig") || target.contains("openbsd") ||
-       target.contains("msvc") || target.contains("emscripten") || target.contains("fuchsia") {
+       target.contains("msvc") || target.contains("emscripten") || target.contains("fuchsia") ||
+       target.contains("redox") {
```
We do not support jemalloc, so we will build with the dummy.

jackpot51 commented on src/liballoc_system/lib.rs:
```diff
@@ -71,7 +72,49 @@ pub extern "C" fn __rust_usable_size(size: usize, align: usize) -> usize {
 
-#[cfg(unix)]
+#[cfg(target_os = "redox")]
+mod imp {
```
Just so we have a stub for potentially using a `libc` allocator, we implement `alloc_system` similarly to how the `unix` implementation works. If desired, these two imp blocks could be combined.

jackpot51 replied:
I will probably change this to be different from Unix so that it defines externs that ralloc will fulfill.

jackpot51 commented on src/liballoc_system/lib.rs:
```diff
@@ -71,7 +72,49 @@ pub extern "C" fn __rust_usable_size(size: usize, align: usize) -> usize {
+    use MIN_ALIGN;
+
+    pub unsafe fn allocate(size: usize, _align: usize) -> *mut u8 {
+        libc::malloc(size as libc::size_t) as *mut u8
```
memalign is not used like in the `unix` implementation, perhaps this is a problem

alexcrichton replied:
Yes I this is one case where I think you'll want to align with the implementation above. This really ends up just needing an allocator interface which supports alignment as an argument, and that's what `memalign` is targeted at.

jackpot51 replied:
Understood, thanks

jackpot51 commented on src/liballoc_system/lib.rs:
```diff
@@ -19,6 +19,7 @@
             issue = "27783")]
 #![feature(allocator)]
 #![feature(staged_api)]
+#![cfg_attr(target_os = "redox", feature(libc))]
```
Note that `ralloc` is used as the default allocator in `librust_back`. What we have done here is to ensure that `liballoc_system` builds, as it is a required dependency of `libstd`. We never link to it, however.

jackpot51 commented on src/librustc_back/target/mod.rs:
```diff
@@ -69,6 +69,7 @@ mod windows_base;
 mod windows_msvc_base;
 mod thumb_base;
 mod fuchsia_base;
+mod redox_base;
```
We must add a new OS base for Redox

jackpot51 commented on src/librustc_back/target/mod.rs:
```diff
@@ -184,6 +185,8 @@ supported_targets! {
     ("aarch64-unknown-fuchsia", aarch64_unknown_fuchsia),
     ("x86_64-unknown-fuchsia", x86_64_unknown_fuchsia),
 
+    ("x86_64-unknown-redox", x86_64_unknown_redox),
```
We support a single architecture for Redox, `x86_64`

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+use target::TargetOptions;
+use std::default::Default;
+
+pub fn opts() -> TargetOptions {
```
This target is a bit different from others. Redox does not link to a `libc`, and does not have a rational C cross compiler. For this reason, some things have to be added. I will explain later.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+            "-Wl,-z,noexecstack".to_string(),
+
+            // Do not link libc
+            "-nostdlib".to_string(),
```
`-nostdlib` must be added to prevent gcc from attempting to link the host's libc.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+            "-nostdlib".to_string(),
+
+            // Static link
+            "-static".to_string()
```
Redox does not support dynamic linking, so all files should be linked statically.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        ],
+        late_link_args: vec![
+            // Link to openlibm for math functions
+            "-lopenlibm".to_string()
```
The only C library we utilize is `openlibm` for math functions, we must specify it explicitly. It is compiled out of tree from here: https://github.com/redox-os/openlibm

alexcrichton replied:
In general the arguments passed in this file, `src/librustc_back/target/*.rs` are linker arguments required for *all* binaries generated for the target. Presumably, though, openlibm isn't required for all binaries? 

What I'd recommend here is moving openlibm linkage to the standard library's build script, as this is essentially a dependency of the standard library, not all executables.

jackpot51 replied:
True, it is only a dependency of libstd, not of a #[no_std] binary. I will move this.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        ],
+        executables: true,
+        relocation_model: "static".to_string(),
+        disable_redzone: true,
```
Red zones are not currently supported by Redox. This may be enabled in the future after testing.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+            "-lopenlibm".to_string()
+        ],
+        executables: true,
+        relocation_model: "static".to_string(),
```
There is no support for relocation in Redox.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+            // Link to openlibm for math functions
+            "-lopenlibm".to_string()
+        ],
+        executables: true,
```
We are able to generate executables. This is done rather unhygienically in `libstd` with a `_start` function.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        executables: true,
+        relocation_model: "static".to_string(),
+        disable_redzone: true,
+        eliminate_frame_pointer: false,
```
We do not eliminate the frame pointer so that stack frames are easily traversable.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        relocation_model: "static".to_string(),
+        disable_redzone: true,
+        eliminate_frame_pointer: false,
+        target_family: Some("redox".to_string()),
```
This is a controversial change. With the way things are currently handled, setting this to `None` implies `Some("unix".to_string())`, which enables a lot of code that is not compatible with Redox.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        eliminate_frame_pointer: false,
+        target_family: Some("redox".to_string()),
+        linker_is_gnu: true,
+        no_default_libraries: true,
```
No default libraries must be linked, as they would come from the host.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        target_family: Some("redox".to_string()),
+        linker_is_gnu: true,
+        no_default_libraries: true,
+        lib_allocation_crate: "ralloc".to_string(),
```
`ralloc` is used as the default allocation crate. It is meant to be compiled out of tree from this: https://github.com/redox-os/ralloc/tree/redox_cross

alexcrichton replied:
I think we may have discussed this a bit on IRC, but I think this'll be best as `alloc_system` as these values are really intended for crates which are shipped as part of the main distribution. If this is a crate which is supplied externally it could cause lots of problems down the road.

jackpot51 replied:
I will set it back to alloc_system and have alloc_system use externs that ralloc will fill in

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        no_default_libraries: true,
+        lib_allocation_crate: "ralloc".to_string(),
+        exe_allocation_crate: "ralloc".to_string(),
+        has_elf_tls: true,
```
We do support TLS, so we enable it.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,53 @@
+        lib_allocation_crate: "ralloc".to_string(),
+        exe_allocation_crate: "ralloc".to_string(),
+        has_elf_tls: true,
+        panic_strategy: PanicStrategy::Abort,
```
We do not support unwinding, so we panic using `abort`

jackpot51 commented on src/librustc_back/target/x86_64_unknown_redox.rs:
```diff
@@ -0,0 +1,30 @@
+
+use target::{Target, TargetResult};
+
+pub fn target() -> TargetResult {
```
These settings are very similar to those in other targets, and are nothing special.

jackpot51 commented on src/libstd/build.rs:
```diff
@@ -26,7 +26,7 @@ fn main() {
     let host = env::var("HOST").expect("HOST was not set");
     if cfg!(feature = "backtrace") && !target.contains("apple") && !target.contains("msvc") &&
-        !target.contains("emscripten") && !target.contains("fuchsia") {
+        !target.contains("emscripten") && !target.contains("fuchsia") && !target.contains("redox") {
```
We disable the build of `libbacktrace`, as it will not currently compile for Redox

jackpot51 commented on src/libstd/lib.rs:
```diff
@@ -299,6 +299,7 @@
 #![feature(unwind_attributes)]
 #![feature(vec_push_all)]
 #![feature(zero_one)]
+#![cfg_attr(target_os = "redox", feature(naked_functions))]
```
#### Hygiene error! In order to generate executables, we have to have a naked `_start` function at the libstd root.

jackpot51 commented on src/libstd/lib.rs:
```diff
@@ -422,6 +423,11 @@ pub use core_collections::vec;
+// Reexport the start module on platforms that provide it
+#[unstable(feature = "start_fn", issue="0")]
+#[cfg(target_os = "redox")]
+pub use sys::start::*;
```
#### Hygiene error! In order to generate executables, we have to have a naked `_start` function at the libstd root.

alexcrichton replied:
We actually currently have `src/libstd/rt.rs` which is full of hacks like this, could you move this over there?

jackpot51 replied:
The thing is that it must be exposed at the crate root. I could have pub use sys::start; in libstd::rt, but then rt::start::* would still have to be pub used in lib.rs

alexcrichton replied:
I believe you just need the symbols exported, right? In that case they just need to be reachable publicly, not literally at the crate root. That's what the `rt` module suffices to do (it's entirely unstable)

jackpot51 replied:
Ah, yes.

jackpot51 commented on src/libstd/sys/redox/mod.rs:
```diff
@@ -10,8 +10,6 @@
 
 #![allow(dead_code, missing_docs, bad_style)]
 
-pub extern crate syscall;
```
We have moved the `syscall` crate in-tree to the `syscall` module.

jackpot51 commented on src/libstd/sys/redox/net/mod.rs:
```diff
@@ -20,10 +20,11 @@ use vec::{IntoIter, Vec};
 
 use self::dns::{Dns, DnsQuery};
 
-pub extern crate libc as netc;
```
We no longer use the `libc` crate for `netc` definitions. They are now inside the `netc` module.

jackpot51 commented on src/libstd/sys/redox/net/netc.rs:
```diff
@@ -0,0 +1,47 @@
+pub type in_addr_t = u32;
```
These are the definitions from `libc` that are relevant in `std::net`

jackpot51 commented on src/libstd/sys/redox/rand.rs:
```diff
@@ -9,15 +9,19 @@
 use rand::Rng;
 
-pub struct OsRng;
+// FIXME: Use rand:
```
We have to specify a simple `OsRng` algorithm until the `rand:` process is started. If you have more questions, please ask them here. In the future, `rand:` will be in the kernel to allow the `init` process to execute. (Rust currently uses `OsRng` in every process for the `HashState`)

jackpot51 commented on src/libstd/sys/redox/start.rs:
```diff
@@ -0,0 +1,118 @@
+use sys::syscall::exit;
```
#### Hygiene error! In order to generate executables, we have to have a naked `_start` function at the libstd root.

This code would typically be done in `crt0` in `libc`. Since we are avoiding the linkage to a `libc`, we define it here. Due to the way `#[no_mangle]` works, the linker will fail if these functions are not exported all the way to the crate root.

jackpot51 commented on src/libstd/sys/redox/start.rs:
```diff
@@ -0,0 +1,118 @@
+/// Copy N bytes of memory from one location to another.
+#[unstable(feature = "start_fn", issue = "0")]
+#[no_mangle]
+pub unsafe extern fn memcpy(dest: *mut u8, src: *const u8,
```
We also have to define some C-like functions that are used by `llvm` to avoid linker errors.

jackpot51 commented on src/libstd/sys/redox/syscall/call.rs:
```diff
@@ -0,0 +1,287 @@
+use super::arch::*;
```
These are all of the Redox system calls.

jackpot51 commented on src/libstd/sys/redox/syscall/data.rs:
```diff
@@ -0,0 +1,127 @@
+use core::ops::{Deref, DerefMut};
```
These are the data types used in Redox system calls

jackpot51 commented on src/libstd/sys/redox/syscall/flag.rs:
```diff
@@ -0,0 +1,84 @@
+pub const CLONE_VM: usize = 0x100;
```
These are flags used in Redox system calls.

jackpot51 commented on src/libstd/sys/redox/syscall/error.rs:
```diff
@@ -0,0 +1,315 @@
+use core::{fmt, result};
```
This is the definition of `Error`s and `Result`s for Redox system calls

jackpot51 commented on src/libstd/sys/redox/syscall/mod.rs:
```diff
@@ -0,0 +1,33 @@
+pub use self::arch::*;
```
This is the Redox system call module, now in-tree.

jackpot51 commented on src/libstd/sys/redox/syscall/number.rs:
```diff
@@ -0,0 +1,63 @@
+pub const SYS_CLASS: usize =    0xF000_0000;
```
These are the call numbers for each system call.

jackpot51 commented on src/libstd/sys/redox/syscall/scheme.rs:
```diff
@@ -0,0 +1,232 @@
+use core::{mem, slice};
```
This is the definition of a Redox scheme. It could probably be removed.

jackpot51 commented on src/libstd/sys/redox/syscall/x86.rs:
```diff
@@ -0,0 +1,72 @@
+use super::error::{Error, Result};
```
This is the system call ABI definition for x86

jackpot51 commented on src/libstd/sys/redox/syscall/x86_64.rs:
```diff
@@ -0,0 +1,72 @@
+use super::error::{Error, Result};
```
This is the system call ABI definition for x86_64

jackpot51 commented on src/libtest/lib.rs:
```diff
@@ -914,6 +914,11 @@ fn use_color(opts: &TestOpts) -> bool {
 }
 
+#[cfg(target_os = "redox")]
+fn stdout_isatty() -> bool {
```
Redox needs to support `isatty`. We return `false` as a safe alternative

jackpot51 commented on src/libtest/lib.rs:
```diff
@@ -1103,6 +1108,12 @@ fn get_concurrency() -> usize {
     }
 
+    #[cfg(target_os = "redox")]
+    fn num_cpus() -> usize {
```
Redox needs to provide a way to get the number of CPUs. We return `1` as a safe alternative

jackpot51 commented on src/liballoc_system/lib.rs:
```diff
@@ -19,7 +19,7 @@
 #![feature(allocator)]
 #![feature(staged_api)]
-#![cfg_attr(unix, feature(libc))]
+#![cfg_attr(any(unix, target_os = "redox"), feature(libc))]
```
For `alloc_system`, we simply reuse the `unix` setup to allow a `libc` implementation of an allocator

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,49 @@
+        linker_is_gnu: true,
+        no_default_libraries: true,
+        lib_allocation_crate: "alloc_system".to_string(),
+        exe_allocation_crate: "alloc_system".to_string(),
```
We use `alloc_system` as the default allocator. I personally would like to use `ralloc`

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,49 @@
+        lib_allocation_crate: "alloc_system".to_string(),
+        exe_allocation_crate: "alloc_system".to_string(),
+        has_elf_tls: true,
+        panic_strategy: PanicStrategy::Abort,
```
Unwinding is not supported at the moment, so we abort.

jackpot51 commented on src/libstd/rt.rs:
```diff
@@ -22,11 +22,13 @@
 
+// Reexport the start module on platforms that provide it
+#[unstable(feature = "sys_rt", issue="0")]
+pub use sys::rt::*;
```
Hygiene error! In order to generate executables, we have to have a naked _start function at the libstd root.

jackpot51 commented on src/libstd/sys/redox/rt.rs:
```diff
@@ -0,0 +1,130 @@
+#[no_mangle]
+#[naked]
+#[cfg(target_arch = "x86")]
+pub unsafe fn _start() {
```
Hygiene error! In order to generate executables, we have to have a naked _start function at the libstd root.

jackpot51 commented on src/libstd/sys/redox/rt.rs:
```diff
@@ -0,0 +1,130 @@
+#[naked]
+#[cfg(target_arch = "x86_64")]
+pub unsafe fn _start() {
+    asm!("mov rdi, rsp
```
Hygiene error! In order to generate executables, we have to have a naked _start function at the libstd root.

RobertWHurst commented:
Can't wait for this to land 👌🏻

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -0,0 +1,50 @@
+            // Static link
+            "-static".to_string()
+        ],
+        late_link_args: vec![
```
We late link libc and libm so that they can be provided when binaries are linked, rather than using the compiler defaults.

brson commented:
This looks great to me.

The only thing that bothers me here is `target_family`. I still don't want to define a new family. I think we should change the logic in the driver so that None means `target_family` is not defined, and change all the unix target definitions to define it as unix.

jackpot51 commented:
Ok @brson - Please see https://github.com/rust-lang/rust/pull/38401/commits/c59bb4979c6d2e994b01d6ed9e005f8881a36935

I changed all of the Unix platforms to `target_family: Some("unix".to_string())` and the Windows platforms to `Some("windows".to_string())`. I changed Redox to `None` I think I covered all the "bases".

I changed librustc/session/config.rs to accept target_family only from the target configuration, to insert it as `cfg(target_family = "...")` and if it is `windows` or `unix`, to insert it as `cfg(...)`

jackpot51 commented on src/librustc/session/config.rs:
```diff
@@ -943,26 +943,20 @@ pub fn default_configuration(sess: &Session) -> ast::CrateConfig {
 
-    let fam = if let Some(ref fam) = sess.target.target.options.target_family {
-        Symbol::intern(fam)
-    } else if sess.target.target.options.is_like_windows {
```
Since the windows platforms now correctly define target_family, this is no longer necessary.

jackpot51 commented on src/librustc_back/target/redox_base.rs:
```diff
@@ -38,7 +38,7 @@ pub fn opts() -> TargetOptions {
         disable_redzone: true,
         eliminate_frame_pointer: false,
-        target_family: Some("redox".to_string()),
+        target_family: None,
```
Redox no longer has a family :(

brson replied:
😢

jackpot51 commented on src/librustc_back/target/windows_base.rs:
```diff
@@ -24,6 +24,7 @@ pub fn opts() -> TargetOptions {
         staticlib_prefix: "".to_string(),
         staticlib_suffix: ".lib".to_string(),
         no_default_libraries: true,
+        target_family: Some("windows".to_string()),
```
Both the windows bases are covered

jackpot51 commented on src/librustc_back/target/apple_base.rs:
```diff
@@ -37,6 +37,7 @@ pub fn opts() -> TargetOptions {
         function_sections: false,
         dynamic_linking: true,
         executables: true,
+        target_family: Some("unix".to_string()),
```
Everything that isn't windows, is set to Unix. That includes Linux, OS X, iOS, NetBSD, OpenBSD, FreeBSD, Bitrig, Solaris, and Haiku

brson commented:
@bors r+

I'm very happy with how this is turning out so far. I wish there wasn't so much cfg sprawl in non-std crates, but that's something we can figure out in the future.

Tagging with relnotes not only for the redox support, but also the breaking change to target_family.

bors commented:
:pushpin: Commit 4dcb867 has been approved by `brson`

<!-- @bors r=brson 4dcb86767111149bcd233d6ac5c782d345d1e10b -->

bors commented:
:hourglass: Testing commit 4dcb867 with merge 99913c5...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/11380), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/11445), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/2667), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/2486), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/4654), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/11387), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/11333), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/3311), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/11306), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/4537), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/6484), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/11431), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/11404), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/3348), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/2529), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/6530), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/3381), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/6493), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/5677), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2689), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/6505), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/3368)
Approved by: brson
Pushing 99913c5ead4d4f82ecda525b5d6f25dfaf541517 to master...

jackpot51 commented:
Thanks @brson
